### DISCUSSION TITLE
Notice of impending Node baseline update

### DISCUSSION THREAD

**Discussion by bretambrose** [awslabs/aws-crt-nodejs#468] [Announcements]
This thread is intended to serve as a discussion point for a pending update to the Node baseline for both the [AWS JS CRT](https://github.com/awslabs/aws-crt-nodejs) and the [AWS v2 IoT SDK](https://github.com/aws/aws-iot-device-sdk-js-v2) for Javascript.  

Currently the CRT's minimum supported version of Node is 10.16.  After this update, the minimum supported version will move to 14.x.  This update will trigger the following changes:

* Node 10 and 12 will no longer be supported by either the [v2 IoT SDK](https://github.com/aws/aws-iot-device-sdk-js-v2) or [CRT](https://github.com/awslabs/aws-crt-nodejs)
* Repository READMEs will be updated to reflect the new baseline going forward and give guidance on version compatibility.
* Internal CI jobs that use 10 or 12 will be upgraded if possible to 14; jobs that cannot be upgrade will be dropped or reworked.
* Dependencies that have been purposely held back due to compatibility issues (cmake-js, axios) will be updated appropriately.
* (Tentative Proposal) To better guide future decision-making, the default [v2 IoT SDK](https://github.com/aws/aws-iot-device-sdk-js-v2) metrics string (as specified in the MQTT username property of the CONNECT packet) will be updated to include Node or browser version info.

No API changes will be made as part of this update.

We are currently planning for this change to be a minor version bump.  We do not yet have a concrete timeline for when the update will take place, but anticipate waiting at least several months for feedback before doing so.

**Comment by xmariopereira** [comment]
Hi, 

In AWS documentation now states: 

```
Your Lambda function must use one of the following runtimes:

Python 3.9 – python3.9

Python 3.8 – python3.8

Python 3.7 – python3.7

Python 2.7 – python2.7 *

Java 11 – java11

Java 8 – java8

Node.js 18 – nodejs18.x

Node.js 16 – nodejs16.x

Node.js 14 – nodejs14.x

Node.js 12 – nodejs12.x *

Node.js 10 – nodejs10.x *
```

source: https://docs.aws.amazon.com/greengrass/v2/developerguide/run-lambda-functions.html

But in https://github.com/aws/aws-iot-device-sdk-js-v2/tree/main, says:
```
Impending Node Version Update
In the coming months, the v2 IoT SDK will be updating its Node baseline from 10.16 to 14. A https://github.com/awslabs/aws-crt-nodejs/discussions/468 has been created for any questions or feedback you may have. We do not yet have a concrete timeline for when the update will happen.
```

So will aws-iot-device-sdk-js-v2 work with node 18?

**Comment by bretambrose** [comment]
Per the other discussion, yes.

**Comment by ashishpandey001** [comment]
considering that node 14 has already reached its end of life and AWS won't patch node 14 runtime with security patches and other updates from 27/11/2023, we should be targeting node 16 instead of node 14 for this update.

**Reply by bretambrose** [reply]
Repeating a comment I just made elsewhere, I don't think middleware libraries should be making minimum node version decisions (after initial release) for any reason other than true need.  We are only doing this because our downstream dependencies have forced this on us.  

Whether or not a given version of node is getting security patches should not have any connection to what versions of node you can use the library on.

I have started the upgrade-to-14 process, but it may take a while because we have some node-12-only CI jobs that are important that I'll need to find a replacement for.

**Comment by bretambrose** [comment]
This update has been completed as of [v1.19.0](https://github.com/awslabs/aws-crt-nodejs/releases/tag/v1.19.0)
